docs/v1/P4Runtime-Spec.mdk

```diff
@@ -1016,6 +1016,104 @@ table my_ipv4_lkup {
   ...
 }
 ~ End p4example
+### Custom Annotations in P4Info
+P4 allows flexible annotations[@P4Annotations]. Users may add their own annotations as long as they obey the language syntax. To facilitate consumption of user-defined annotations, P4Info has an `annotation_map` for any element which can accept such annotations, &ie; the `PkgInfo`, `Preamble`, `MatchField`, `ActionRef`, `Action` and `ControllerPacketMetadata` messages. The `annotation_map` contains a suitable parsed representation for the different types of annotation bodies, namely: `expression_list` and `kv_map`. Empty annotation bodies are signified by an empty `value` field  in the `annotation_map` entry. 
+
+Note that `AnnotationBody` allows for simultaneous `expression_list` and `kv_map` entries even though the P6-16 specification grammar does not at this writing.
+
+~ Begin Proto
+// A parsed representation of the annotation body.
+// Ex. empty annotation:           @my_anno1()
+// Ex. expression list annotation: @my_anno2(1+1, "Hello", 1 << 3)
+// Ex. KV-pair annotation:         @my_anno3(a=b,c=d)
+// Ex. Cumulative kv-pair          @my_anno3(e=f)  // adds to my_anno3's properties
+message AnnotationBody {
+  repeated string expressions = 1;
+  map <string,string> properties = 2;
+}
+~ End Proto
+
+Given the following P4 source fragment:
+~ Begin P4Example
+@my_anno1()
+@my_anno2("Some random content: for i in range(10)")
+@my_anno3(1+1, "Hello", 1 << 3)
+@my_anno4(a=b,c=d)
+@my_anno4(e=f)
+@my_anno5(w=x, "expression one", y=z, expression two)
+table test1 {
+    key = { hdrs.data.f1 : ternary; }
+    actions = {
+        setb1;
+        noop;
+    }
+    default_action = noop;
+}
+~ End P4Example
+The corresponding P4Info `Preamble` fragment would be as follows. Note the appearance of the raw `annotations` fields which are deprecated in favor of the `annotation_map` fields.
+~ Begin Prototext
+  preamble {
```

Comment:
this block is still indented for some reason?

Reply:
I added the missing `table {` opening line which now makes the indent sensible.